WalletWasabi/BitcoinCore/CachedRpcClient.cs

```diff
@@ -0,0 +1,209 @@
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Primitives;
+using NBitcoin;
+using NBitcoin.RPC;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace WalletWasabi.BitcoinCore
+{
+	public class CachedRpcClient : RpcClientBase
+	{
+		private CancellationTokenSource _tipChangeCancellationTokenSource;
+
+		public CachedRpcClient(RPCClient rpc, IMemoryCache cache)
+		: base(rpc)
+		{
+			Cache = cache;
+		}
+		
+		public IMemoryCache Cache { get; }
+
+		private CancellationTokenSource TipChangeCancellationTokenSource
+		{
+			get 
+			{
+				if (_tipChangeCancellationTokenSource is null || _tipChangeCancellationTokenSource.IsCancellationRequested)
+				{
+					_tipChangeCancellationTokenSource = new CancellationTokenSource();
+				}
+				return _tipChangeCancellationTokenSource;
+			}
+		}
+
+		public override async Task<uint256> GetBestBlockHashAsync()
+		{
+			string cacheKey = nameof(GetBestBlockHashAsync);
+			if (!Cache.TryGetValue(cacheKey, out uint256 blockHash))
+			{
+				blockHash = await base.GetBestBlockHashAsync().ConfigureAwait(false);
+
+				var cacheEntryOptions = new MemoryCacheEntryOptions()
+					.SetSize(1)
```

Comment:
I assume this is size of 1 because its a single entry?

Reply:
The `MemoryCache` uses a size unit that is not tight to the real entries' size in bytes. In this way every item has a weight in the cache independently of how much memory it really uses. So, here `1` means that I want to keep as many of these entries in memory (I will be able to store up to 1000 of these elements)